infra/Pos/Slotting/MemState/Holder.hs

```diff
@@ -42,22 +51,72 @@ class HasSlottingVar ctx where
 type SlotsDefaultEnv ctx m =
     (MonadReader ctx m, HasSlottingVar ctx, MonadIO m)
 
-getSystemStartDefault :: SlotsDefaultEnv ctx m => m Timestamp
+getSystemStartDefault
+    :: SlotsDefaultEnv ctx m
+    => m Timestamp
 getSystemStartDefault = view slottingTimestamp
 
-getSlottingDataDefault :: SlotsDefaultEnv ctx m => m SlottingData
-getSlottingDataDefault = atomically . readTVar =<< view slottingVar
+getAllEpochIndicesDefault
+    :: SlotsDefaultEnv ctx m
+    => m [EpochIndex]
+getAllEpochIndicesDefault = do
+    var <- view slottingVar
+    atomically $ getAllEpochIndices <$> readTVar var
+
+getCurrentEpochIndexDefault
+    :: SlotsDefaultEnv ctx m
+    => m EpochIndex
+getCurrentEpochIndexDefault = do
+    var <- view slottingVar
+    atomically $ getCurrentEpochIndex <$> readTVar var
+
+getNextEpochIndexDefault
+    :: SlotsDefaultEnv ctx m
+    => m EpochIndex
+getNextEpochIndexDefault = do
+    var <- view slottingVar
+    atomically $ getNextEpochIndex <$> readTVar var
+
+getCurrentEpochSlottingDataDefault
+    :: SlotsDefaultEnv ctx m
+    => m EpochSlottingData
+getCurrentEpochSlottingDataDefault = do
+    var <- view slottingVar
+    atomically $ getCurrentEpochSlottingData <$> readTVar var
+
+getNextEpochSlottingDataDefault
+    :: SlotsDefaultEnv ctx m
+    => m EpochSlottingData
+getNextEpochSlottingDataDefault = do
+    var <- view slottingVar
+    atomically $ getNextEpochSlottingData <$> readTVar var
 
-waitPenultEpochEqualsDefault :: SlotsDefaultEnv ctx m => EpochIndex -> m ()
-waitPenultEpochEqualsDefault target = do
+getEpochSlottingDataDefault
+    :: SlotsDefaultEnv ctx m
+    => EpochIndex
+    -> m (Maybe EpochSlottingData)
+getEpochSlottingDataDefault ei = do
+    var <- view slottingVar
+    atomically $ lookupEpochSlottingData ei <$> readTVar var
+
+putEpochSlottingDataDefault
+    :: SlotsDefaultEnv ctx m
+    => EpochIndex
+    -> EpochSlottingData
+    -> m ()
+putEpochSlottingDataDefault ei esd = do
     var <- view slottingVar
     atomically $ do
-        penultEpoch <- sdPenultEpoch <$> readTVar var
-        when (penultEpoch /= target) retry
+        slottingData <- readTVar var
+        writeTVar var (addEpochSlottingData ei esd slottingData)
 
-putSlottingDataDefault :: SlotsDefaultEnv ctx m => SlottingData -> m ()
-putSlottingDataDefault sd = do
+waitCurrentEpochEqualsDefault
+    :: SlotsDefaultEnv ctx m
+    => EpochIndex
+    -> m ()
+waitCurrentEpochEqualsDefault target = do
     var <- view slottingVar
     atomically $ do
-        penultEpoch <- sdPenultEpoch <$> readTVar var
-        when (penultEpoch < sdPenultEpoch sd) $ writeTVar var sd
+        currentEpoch <- getCurrentEpochIndex <$> readTVar var
+        when (currentEpoch /= target) retry
```

Comment:
@pva701 this infinite loop won't consume CPU, but I think you are right.
I propose to change this function to wait until current epoch is ≥ target. @ksaric can you do it?